cider/include/cider/CiderException.h

```diff
@@ -46,15 +48,73 @@ class CiderRuntimeException : public CiderException {
  public:
   explicit CiderRuntimeException(const std::string& msg)
       : CiderException("CiderRuntimeException", msg) {}
-  virtual ~CiderRuntimeException() noexcept {};
+
+  // for subclass inheritance
+  explicit CiderRuntimeException(std::string type, const std::string& msg)
+      : CiderException(type, msg) {}
+
+  virtual ~CiderRuntimeException() noexcept {}
 };
 
 // please use CIDER_THROW(CiderCompileException, "xxx");
 class CiderCompileException : public CiderException {
  public:
   explicit CiderCompileException(const std::string& msg)
       : CiderException("CiderCompileException", msg) {}
-  virtual ~CiderCompileException() noexcept {};
+
+  // for subclass inheritance
+  explicit CiderCompileException(std::string type, const std::string& msg)
+      : CiderException(type, msg) {}
+
+  virtual ~CiderCompileException() noexcept {}
+};
+
+// please use CIDER_THROW(CiderOutOfMemoryException, "xxx");
+class CiderOutOfMemoryException : public CiderRuntimeException {
+ public:
+  explicit CiderOutOfMemoryException(const std::string& msg)
+      : CiderRuntimeException("CiderOutOfMemoryException", msg) {}
+
+  virtual ~CiderOutOfMemoryException() noexcept {}
+};
+
+// please use CIDER_THROW(CiderWatchdogException, "xxx");
+class CiderWatchdogException : public CiderCompileException {
+ public:
+  explicit CiderWatchdogException(const std::string& msg)
+      : CiderCompileException("CiderWatchdogException", msg) {}
+
+  virtual ~CiderWatchdogException() noexcept {}
+};
+
+class CiderHashJoinException : public CiderCompileException {
+ public:
+  explicit CiderHashJoinException(const std::string& msg)
+      : CiderCompileException("CiderHashJoinException", msg) {}
+
+  // for subclass inheritance
+  explicit CiderHashJoinException(std::string type, const std::string& msg)
+      : CiderCompileException(type, msg) {}
+
+  virtual ~CiderHashJoinException() noexcept {}
+};
+
+// please use CIDER_THROW(CiderOneToMoreHashException, "xxx");
+class CiderOneToMoreHashException : public CiderHashJoinException {
```

Comment:
What's the BKM for exception type creation? In another word, when should we create a new exception? Please add comments around design principle here.

Reply:
There is a catch CiderOneToMoreHashException in the code with special handling.